### PR TITLE
editoast: fix unique constraint error for infra cloning

### LINKED ISSUE #5946: editoast: error when cloning an infra, with the error hash unique constraint
### What happened?

When we try to clone an infra, we have an error with the error hash unique constraint.

### What did you expect to happen?

The error should be clone succesfully

### How can we reproduce it (as minimally and precisely as possible)?

1. Go to the editor
2. try to clone an infra

### What operating system, browser and environment are you using?

- **Browser**: Firefox v120
- **OS**: Wondows 11
- **Env**: Dgex dev

### OSRD version (top right corner `Account` button > `Informations`)

7796e6d

### PR DESCRIPTION
**Pull request by Tguisnet**
close #5946

### CODE REVIEW

**Review by flomonster — APPROVED** [review]
Good catch. Can you please add a test?

File: editoast/src/views/infra/mod.rs [hunk 1]
```diff
@@ -561,6 +565,62 @@ pub mod tests {
         assert!(Infra::delete(db_pool, cloned_infra_id).await.unwrap());
     }
 
+    #[rstest]
+    async fn infra_with_errors_clone(
+        db_pool: Data<DbPool>,
+        #[future] empty_infra: TestFixture<Infra>,
+    ) {
```
Comment (flomonster) [inline editoast/src/views/infra/mod.rs]: Why not use a filled infra like `small_infra`. You will have to generate layers but you can check that all objects are well cloned.
This will test more things with fewer manual things to do.

**Review by flomonster — CHANGES_REQUESTED** [review]

File: editoast/src/views/infra/mod.rs [hunk 2]
```diff
@@ -561,6 +565,79 @@ pub mod tests {
...
+            .await
+            .unwrap();
+
+        let errors_count =
+            sql_query("SELECT COUNT (*) as nb from infra_layer_error where infra_id = $1")
+                .bind::<BigInt, _>(small_infra_id)
+                .get_result::<Count>(&mut conn)
+                .await
+                .unwrap();
+        assert!(errors_count.nb > 1);
```
Comment (flomonster) [inline editoast/src/views/infra/mod.rs]: You never check that the clone infra duplicate errors. We don't need to make a special case for errors we can include it in the rest of the tables (see my following comments).

File: editoast/src/views/infra/mod.rs [hunk 3]
```diff
@@ -561,6 +565,79 @@ pub mod tests {
...
+        let cloned_infra_id: i64 = read_body_json(response).await;
+
+        for object in ObjectType::iter() {
+            let model_table = object.get_table();
+
+            let mut vec_table = vec![model_table];
+
+            if let Some(layer_table) = object.get_geometry_layer_table() {
+                vec_table.push(layer_table);
+            }
```
Comment (flomonster) [inline editoast/src/views/infra/mod.rs]: I would split the table list construction and the check of duplication.

```suggestion
       let mut tables = vec!["infra_layer_error"];
        for object in ObjectType::iter() {
            tables.push(object.get_table());
            if let Some(layer_table) = object.get_geometry_layer_table() {
                tables.push(layer_table);
            }
        }
```

File: editoast/src/views/infra/mod.rs [hunk 4]
```diff
@@ -561,6 +565,79 @@ pub mod tests {
...
+
+            for table in vec_table {
+                let mut reference = 0;
+                for inf_id in [small_infra_id, cloned_infra_id] {
```
Comment (flomonster) [inline editoast/src/views/infra/mod.rs]: Instead of using the `reference` variable, we can store counts in a hashmap like `HashMap<TableName, Vec<usize>>`. Once the hashmap is created we can iterate on it to check the consistency and also check that all values are greater than 1.

**Review by flomonster — CHANGES_REQUESTED** [review]

File: editoast/src/views/infra/mod.rs (hunk 2 above)
Comment (Tguisnet) [reply editoast/src/views/infra/mod.rs]: I agree with your following comments, but for the error layer, I still want to have a special case.
I think that we don't need to check the number of objects in our model / layer tables (if you want to do this, maybe we should see a test for the generation of small infra and the generation of the layers) 
The only reason I check that errors are more than 1 is to check the corner case of the new hash for the errors.
If this test failed, we avoid the useless cloning of the infra.
Btw, if I don't make a special case for the error I will at least make a special case for the extended switch type object because its table contains 0 element.

File: editoast/src/views/infra/mod.rs (hunk 4 above)
Comment (Tguisnet) [reply editoast/src/views/infra/mod.rs]: Agree for the idee of the hashmap, but not the second part of check that all values are greater than 1 (see my comment above)

File: editoast/src/views/infra/mod.rs [hunk 5]
```diff
@@ -561,6 +565,91 @@ pub mod tests {
...
+                .unwrap();
+
+                table_content
+                    .entry(table)
+                    .or_insert_with(Vec::new)
+                    .push(count_object.nb);
+            }
+        }
+
+        for (_, val) in table_content.iter() {
```
Comment (flomonster) [inline editoast/src/views/infra/mod.rs]:
```suggestion
        for val in table_content.values() {
```

File: editoast/src/views/infra/mod.rs [hunk 6]
```diff
@@ -561,6 +565,91 @@ pub mod tests {
...
+            assert_eq!(val.len(), 2);
```
Comment (flomonster) [inline editoast/src/views/infra/mod.rs:646]: We can add a comment to explain what is checked here.

**Review by flomonster — APPROVED** [review]
